### ISSUE TITLE
Getting a unit's public_address is unreliable in Juju 2.9

### ISSUE THREAD

**Issue by ChrisMacNaughton** [juju/python-libjuju#551]
Calling [unit.public_address](https://github.com/juju/python-libjuju/blob/master/juju/unit.py#L66-L71) works reliably on juju 2.8 but fails intermittently, but frequently, on juju 2.9.

This method calls into the safe_data attribute on the object which, when successful, looks like:

```python
{'agent-status': {'current': 'idle',
                  'message': '',
                  'since': '2021-09-30T10:05:27.491219135Z',
                  'version': '2.8.10'},
 'application': 'keystone',
 'charm-url': 'cs:~openstack-charmers-next/keystone-563',
 'life': 'alive',
 'machine-id': '1',
 'model-uuid': '5beccf88-9f05-402c-8c97-584b490dceab',
 'name': 'keystone/0',
 'port-ranges': [{'from-port': 5000, 'protocol': 'tcp', 'to-port': 5000}],
 'ports': [{'number': 5000, 'protocol': 'tcp'}],
 'principal': '',
 'private-address': '10.5.0.141',
 'public-address': '10.5.0.141',
 'series': 'bionic',
 'subordinate': False,
 'workload-status': {'current': 'active',
                     'message': 'Unit is ready',
                     'since': '2021-09-30T09:48:11.443378654Z',
                     'version': ''}}
```

As the method is defined as a `@property`, the moment it returns None, it will never re-evaluate, meaning that the unit will _never_ have a public address.

**Comment by ajkavanagh** [comment]
> As the method is defined as a `@property`, the moment it returns None, it will never re-evaluate, meaning that the unit will _never_ have a public address.

I think you're thinking of the `@cached` decorator?  `@property` just means it doesn't need to be 'called' just 'accessed'.

**Comment by ChrisMacNaughton** [comment]
Could be worse then :-)

**Comment by ChrisMacNaughton** [comment]
As noted by Alex, I was incorrect about the @property caching; the issue is that we see the public_address being None well into a deployment, when trying to identify the address of a unit to talk to, but after an entire deployment has stabilized.

**Comment by SimonRichardson** [comment]
It depends if we've got the delta from the `AllWatcher`, or if we're missing something, which could be a possibility. Alternatively, we could make a direct request.

**Comment by cderici** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Getting a unit's public-address in libjuju](https://trello.com/c/ijdPEFa7/4136-getting-a-units-public-address-in-libjuju)

**Comment by ChrisMacNaughton** [comment]
The new function doesn't resolve the issue, as the public_address never seems to show up in this use case.

I deployed 10 units of the ubuntu charm (forcing it to use impish because of another test) and then waited for the model to finish:

```
Model  Controller        Cloud/Region             Version  SLA          Timestamp
thing  icey-serverstack  serverstack/serverstack  2.9.18   unsupported  15:39:33Z

App     Version  Status  Scale  Charm   Store       Rev  OS      Notes
ubuntu           active     11  ubuntu  jujucharms   18  ubuntu  

Unit       Workload  Agent  Machine  Public address  Ports  Message
ubuntu/0*  active    idle   0        10.5.1.64              
ubuntu/1   active    idle   1        10.5.1.58              
ubuntu/2   active    idle   2        10.5.3.44              
ubuntu/3   active    idle   3        10.5.1.127             
ubuntu/4   active    idle   4        10.5.0.237             
ubuntu/5   active    idle   5        10.5.1.48              
ubuntu/6   active    idle   6        10.5.2.135             
ubuntu/7   active    idle   7        10.5.1.90              
ubuntu/8   active    idle   8        10.5.0.14              
ubuntu/9   active    idle   9        10.5.2.66              
ubuntu/10  active    idle   10       10.5.0.23              

Machine  State    DNS         Inst id                               Series  AZ    Message
0        started  10.5.1.64   b3475e8e-aff2-484e-b207-8d37f5be1e34  impish  nova  ACTIVE
1        started  10.5.1.58   a1eca488-27e8-4776-be2a-44a30d726443  impish  nova  ACTIVE
2        started  10.5.3.44   265067b8-2918-4129-8dc9-0ea955e4af51  impish  nova  ACTIVE
3        started  10.5.1.127  82578dfb-7c6a-4e67-8efe-c6d07048de5b  impish  nova  ACTIVE
4        started  10.5.0.237  3b4a25c6-a8c8-41e0-9d7e-7a16dac3f0f1  impish  nova  ACTIVE
5        started  10.5.1.48   e76f4323-70bf-4020-a8b4-372efc838bc6  impish  nova  ACTIVE
6        started  10.5.2.135  04b613a3-6eca-4a62-8f1c-cb1d7effff0f  impish  nova  ACTIVE
7        started  10.5.1.90   1e98467a-17f2-4ef2-9c9a-b3d733ec7d88  impish  nova  ACTIVE
8        started  10.5.0.14   946c384d-c3fc-4fe1-a3bf-23c596c954d6  impish  nova  ACTIVE
9        started  10.5.2.66   19a91663-3d44-48bb-a056-17ef4dc620d7  impish  nova  ACTIVE
10       started  10.5.0.23   5f0e79f5-a5b2-4db3-84f5-dace9567c8cd  impish  nova  ACTIVE
```

I then (10+ minutes later) ran the following python code:

```python3
from juju.model import Model
import asyncio

async def do_the_thing():
    model = Model()
    await model.connect_model('thing')
    for unit in model.applications['ubuntu'].units:
      print("{} Address: {}".format(unit.name, await unit.get_public_address()))

loop = asyncio.get_event_loop()
loop.run_until_complete(do_the_thing())
```

which outputs:

```
Task exception was never retrieved
future: <Task finished coro=<Connection._pinger() done, defined at /home/ubuntu/keystone/.tox/func-target/lib/python3.6/site-packages/juju/client/connection.py:502> exception=Exception('No facade Pinger in facades {}',)>
Traceback (most recent call last):
  File "/home/ubuntu/keystone/.tox/func-target/lib/python3.6/site-packages/juju/client/connection.py", line 516, in _pinger
    pinger_facade = client.PingerFacade.from_connection(self)
  File "/home/ubuntu/keystone/.tox/func-target/lib/python3.6/site-packages/juju/client/_client.py", line 67, in from_connection
    connection.facades))
Exception: No facade Pinger in facades {}
unknown facade CAASModelConfigManager
unexpected facade CAASModelConfigManager found, unable to decipher version to use
unknown facade RaftLease
unexpected facade RaftLease found, unable to decipher version to use
unknown facade Secrets
unexpected facade Secrets found, unable to decipher version to use
unknown facade SecretsManager
unexpected facade SecretsManager found, unable to decipher version to use
unknown facade SecretsRotationWatcher
unexpected facade SecretsRotationWatcher found, unable to decipher version to use
ubuntu/0 Address: 10.5.1.64
ubuntu/2 Address: 10.5.3.44
ubuntu/5 Address: None
ubuntu/3 Address: 10.5.1.127
ubuntu/4 Address: None
ubuntu/1 Address: 10.5.1.58
ubuntu/6 Address: 10.5.2.135
ubuntu/7 Address: 10.5.1.90
ubuntu/9 Address: None
ubuntu/10 Address: 10.5.0.23
ubuntu/8 Address: 10.5.0.14
```

It takes a hair over three minutes (yay for 1 minute timeouts) to output that.

In short, the fix landed in #573 doesn't resolve this issue

**Comment by SimonRichardson** [comment]
This is a juju bug, so closing this one now as pylibjuju shouldn't have to change (unless there is really no workaround!). 

https://bugs.launchpad.net/juju/+bug/1950799

**Comment by wolsen** [comment]
Note, PR #600 does not fix this issue. Can still see the problem. See my comment there for more information.

**Comment by SimonRichardson** [comment]
@wolsen the old way is non sustainable, so to correctly do this in pylibjuju so aligns with the latest Juju is to call the API facade. If the value isn't there, you are expected to poll that method. This is the new way to ensure that we have the correct information for all substrates.

If you feel differently then comment on the Juju bug.

**Comment by wolsen** [comment]
> @wolsen the old way is non sustainable, so to correctly do this in pylibjuju so aligns with the latest Juju is to call the API facade. If the value isn't there, you are expected to poll that method. This is the new way to ensure that we have the correct information for all substrates.
> 
> If you feel differently then comment on the Juju bug.

No opinion on the new way vs the old way. IMO, the new way is acceptable, it just doesn't work without the additional patch in PR #610, which you are already aware of.

**Comment by SimonRichardson** [comment]
> No opinion on the new way vs the old way. IMO, the new way is acceptable, it just doesn't work without the additional patch in PR #610, which you are already aware of.

Indeed. On it.
